src/main/java/org/jenkinsci/plugins/gitclient/GitUsernamePasswordBinding.java

```diff
@@ -0,0 +1,158 @@
+        setKeyBindings(credentials);
+        gitTool = gitToolName(run, taskListener);
+        if (gitTool != null && filePath != null) {
+            final UnbindableDir unbindTempDir = UnbindableDir.create(filePath);
+            setRunEnvironmentVariables(filePath, taskListener);
+            GenerateGitScript gitScript = new GenerateGitScript(credentials.getUsername(),
+                    credentials.getPassword().getPlainText(), credentials.getId());
+            FilePath gitTempFile = gitScript.write(credentials, unbindTempDir.getDirPath());
+            credMap.put("GIT_ASKPASS", gitTempFile.getRemote());
+            return new MultiEnvironment(credMap, unbindTempDir.getUnbinder());
+        } else {
+            return new MultiEnvironment(credMap);
+        }
+    }
+
+    @Override
+    public Set<String> variables() {
+        Set<String> keys = new LinkedHashSet<>();
+        keys.add(GIT_USERNAME_KEY);
+        keys.add(GIT_PASSWORD_KEY);
+        return keys;
+    }
+
+
+    @Override
+    public void setKeyBindings(@NonNull StandardCredentials credentials) {
+        credMap.put(GIT_USERNAME_KEY, ((StandardUsernamePasswordCredentials) credentials).getUsername());
+        credMap.put(GIT_PASSWORD_KEY, ((StandardUsernamePasswordCredentials) credentials).getPassword().getPlainText());
+    }
+
+    @Override
+    public void setRunEnvironmentVariables(@NonNull FilePath filePath, @NonNull TaskListener listener) throws IOException, InterruptedException {
+        if (!Functions.isWindows() && ((CliGitAPIImpl) getGitClientInstance(listener)).
+                isAtLeastVersion(2, 3, 0, 0)) {
+            credMap.put("GIT_TERMINAL_PROMPT", "false");
+        } else {
+            credMap.put("GCM_INTERACTIVE", "false");
+        }
+    }
+
+    @Override
+    public GitClient getGitClientInstance(TaskListener listener) throws IOException, InterruptedException {
+        Git gitInstance = Git.with(listener, new EnvVars()).using(gitTool);
+        return gitInstance.getClient();
+    }
+
+    protected static final class GenerateGitScript extends AbstractOnDiskBinding<StandardUsernamePasswordCredentials> {
+
+        private String userVariable = null;
+        private String passVariable = null;
+
+        protected GenerateGitScript(String gitUsername, String gitPassword, String credentialId) {
+            super(gitUsername + ":" + gitPassword, credentialId);
+            this.userVariable = gitUsername;
+            this.passVariable = gitPassword;
+        }
+
+        @Override
+        protected FilePath write(StandardUsernamePasswordCredentials credentials, FilePath workspace)
+                throws IOException, InterruptedException {
+            FilePath gitEcho;
+            String newLine = System.lineSeparator();    //Platform dependent newLine
+            if (!Functions.isWindows()) {
+                gitEcho = workspace.createTempFile("auth", ".sh");
+                gitEcho.write("#!/bin/sh" + newLine +
```

Comment:
should not assume `#!bin/sh`  the shell can be set for each agent and may be different